### PR TITLE
Adding Avi ansible lookup module

### PR DESCRIPTION
**Pull request by sabandi**
Adding new lookup plugin to allow lookups against Avi controller objects

- New Module Pull Request

avi lookup plugin

```paste below

- debug: msg="{{ lookup('avi', avi_credentials=avi_credentials, obj_type='virtualservice') }}"
- debug: msg="{{ lookup('avi', avi_credentials=avi_credentials, obj_name='vs1', obj_type='virtualservice', wantlist=True) }}"
- debug: msg="{{ lookup('avi', obj_uuid='virtualservice-5c0e183a-690a-45d8-8d6f-88c30a52550d', obj_type='virtualservice') }}"
# Using query passes wantlist implicitly
- debug: msg="{{ query('avi', obj_uuid='virtualservice-5c0e183a-690a-45d8-8d6f-88c30a52550d', obj_type='virtualservice') }}"

```

### CODE REVIEW

**Comment by sabandi** [comment]
Hi @chaitanyaavi , @grastogi23 

Could you please review this PR. I need some help what `session_params/get_params` we should skip as mentioned by @grastogi23 in another PR.

Thanks,
Sandeep

File: lib/ansible/plugins/lookup/avi.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            avi = ApiSession(**session_params)
+            path = kwargs['obj_type']
+
+            if kwargs.get('obj_name', None):
+                name = kwargs['obj_name']
+                response = [avi.get_object_by_name(path, name, **get_params)]
+            elif kwargs.get('obj_uuid', None):
+                obj_uuid = kwargs['obj_uuid']
+                obj_path = "%s/%s" % (path, obj_uuid)
+                response = [avi.get(obj_path, **get_params).json()]
```
Comment (grastogi23) [inline lib/ansible/plugins/lookup/avi.py]: we need to check for errors. If the avi.get returns error then what happens? How is the error returned.

File: lib/ansible/plugins/lookup/avi.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            else:
+                response = avi.get(path, **get_params).json().get('results', [])
+
+        except AnsibleParserError:
+            raise AnsibleError("Unable to query the Avi Controller")
```
Comment (grastogi23) [inline lib/ansible/plugins/lookup/avi.py]: Need better error here. There can be other errors like network error, Avi Api error etc. For those what is the error returned.

File: lib/ansible/plugins/lookup/avi.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                'timeout': kwargs.get('timeout', None),
+                'params': kwargs.get('params', None),
+                'api_version': kwargs.get('api_version', None),
+            }
+            avi = ApiSession(**session_params)
+            path = kwargs['obj_type']
+
+            if kwargs.get('obj_name', None):
+                name = kwargs['obj_name']
+                response = [avi.get_object_by_name(path, name, **get_params)]
```
Comment (grastogi23) [inline lib/ansible/plugins/lookup/avi.py]: there is a case when get_object_by_name returns None. we should handle it.

File: lib/ansible/plugins/lookup/avi.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                'token': kwargs.get('token', None),
+                'tenant': kwargs.get('tenant', None),
+                'tenant_uuid': kwargs.get('tenant_uuid', None),
+                'verify': kwargs.get('verify', None),
+                'port': kwargs.get('port', None),
+                'timeout': kwargs.get('timeout', None),
+                'api_version': kwargs.get('api_version', None),
+                'retry_conxn_errors': kwargs.get('retry_conxn_errors', None),
+                'data_log': kwargs.get('data_log', None),
+                'avi_credentials': kwargs.get('avi_credentials', None),
```
Comment (grastogi23) [inline lib/ansible/plugins/lookup/avi.py]: let us only expose avi_credentials as only way to do the lookup as we don't have any legacy code to be compatible with.
Everything else should be done using avi_credentials only.

File: lib/ansible/plugins/lookup/avi.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                'session_id': kwargs.get('session_id', None),
+                'csrftoken': kwargs.get('csrftoken', None),
+                'lazy_authentication': kwargs.get('lazy_authentication', None),
+                'max_api_retries': kwargs.get('max_api_retries', None),
+            }
+            get_params = {
```
Comment (grastogi23) [inline lib/ansible/plugins/lookup/avi.py]: Can we support any parameter that user can send and gets applied. This way parameters like refers_to:xxx will naturally get supported.

**Review by grastogi23 — CHANGES_REQUESTED** [review]

**Comment by grastogi23** [comment]
@sandeepavi can you please run it via ansible sanity. I think there are formatting issues it would not like

**Comment by sabandi** [comment]
sure @grastogi23. The PR is still in works. I pushed the partial code for backup. Can I add new labels to this repo for signifying things like WIP, DO NOT REVIEW etc.

**Comment by sabandi** [comment]
PR is ready for review. Ran ansible sanity and it passed.

**Comment by grastogi23** [comment]
hi let us create pull request to the ansible core...

**Review by grastogi23 — APPROVED** [review]

**Comment by rfrommer** [comment]
If pointed explicitly to help text fields needing review, I remain ever willing to point out mistakes and areas of improvement.

**Pull request closed without merging**